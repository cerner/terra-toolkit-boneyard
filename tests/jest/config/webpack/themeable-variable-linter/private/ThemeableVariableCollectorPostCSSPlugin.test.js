jest.mock('postcss-value-parser');

const postcssValueParser = require('postcss-value-parser');
const ThemeableVariableInformation = require('../../../../../../config/webpack/themeable-variable-linter/private/ThemeableVariableInformation');
const ThemeableVariableCollectorPlugin = require('../../../../../../config/webpack/themeable-variable-linter/private/ThemeableVariableCollectorPostCSSPlugin');

describe('Themeable Variable Collector Plugin', () => {
  it('populates the list of themeable variables', () => {
    const themeableVariableInformation = new ThemeableVariableInformation(['default', 'a', 'b']);
    const plugin = ThemeableVariableCollectorPlugin(themeableVariableInformation);

    const mockVarNode1 = {
      type: 'function',
      value: 'var',
      nodes: [{
        value: '--var1',
      }],
    };

    const mockVarNode2 = {
      type: 'function',
      value: 'var',
      nodes: [{
        value: '--var2',
      }],
    };

    const mockNonVarNode = {
      type: 'function',
      value: 'something other than a var',
      nodes: [{
        value: 'something',
      }],
    };

    const mockNonFunctionNode = {
      type: 'nonfunction',
      value: 'var',
      nodes: [{
        value: 'something',
      }],
    };

    postcssValueParser.mockReturnValue({
      walk: jest
        .fn()
        .mockImplementation(func => {
          func(mockVarNode1);
          func(mockVarNode2);
          func(mockNonVarNode);
          func(mockNonFunctionNode);
        }),
    });

    const mockDeclaration = {
      value: 'declaration value',
      source: {
        start: {
          line: 1,
          column: 1,
        },
        input: {
          origin: () => (
            {
              file: 'file1',
              line: 1,
              column: 1,
            }
          ),
        },
      },
    };
    const mockRoot = {
      walkDecls: jest
        .fn()
        .mockImplementation(func => {
          func(mockDeclaration);
        }),
      walkRules: () => {},
    };

    plugin(mockRoot);

    expect(themeableVariableInformation).toMatchSnapshot();
    expect(postcssValueParser).toBeCalledWith('declaration value');
  });

  it('populates the list of tracked variables', () => {
    const themeableVariableInformation = new ThemeableVariableInformation(['default', 'a', 'b']);
    const plugin = ThemeableVariableCollectorPlugin(themeableVariableInformation);

    const mockDecl1 = {
      prop: '--var1',
      value: 1,
      source: {
        start: {
          line: 1,
          column: 1,
        },
        input: {
          origin: () => (
            {
              file: 'file1',
              line: 1,
              column: 1,
            }
          ),
        },
      },
    };

    const mockDecl2 = {
      prop: '--var2',
      value: 2,
      source: {
        start: {
          line: 1,
          column: 1,
        },
        input: {
          origin: () => (
            {
              file: 'file1',
              line: 1,
              column: 1,
            }
          ),
        },
      },
    };

    const mockDecl2Dup = {
      prop: '--var2',
      value: 3,
      source: {
        start: {
          line: 1,
          column: 1,
        },
        input: {
          origin: () => (
            {
              file: 'file1',
              line: 1,
              column: 1,
            }
          ),
        },
      },
    };

    const mockDecl3 = {
      prop: '--var3',
      value: 3,
      source: {
        start: {
          line: 1,
          column: 1,
        },
        input: {
          origin: () => (
            {
              file: 'file1',
              line: 1,
              column: 1,
            }
          ),
        },
      },
    };

    const mockDecl3NonDup = {
      prop: '--var3',
      value: 3,
      source: {
        start: {
          line: 1,
          column: 1,
        },
        input: {
          origin: () => (
            {
              file: 'file1',
              line: 1,
              column: 1,
            }
          ),
        },
      },
    };

    const mockDefaultNode = {
      walkDecls: jest
        .fn()
        .mockImplementation(func => {
          func(mockDecl1);
          func(mockDecl2);
          func(mockDecl3);
        }),
    };
    const mockANode = {
      walkDecls: jest
        .fn()
        .mockImplementation(func => {
          func(mockDecl1);
        }),
    };
    const mockBNode = {
      walkDecls: jest
        .fn()
        .mockImplementation(func => {
          func(mockDecl1);
          func(mockDecl2);
          func(mockDecl2Dup);
          func(mockDecl3);
          func(mockDecl3NonDup);
        }),
    };

    const mockRoot = {
      walkDecls: () => {},
      walkRules: jest
        .fn()
        .mockImplementation((regex, func) => {
          if (String(regex) === String(RegExp('.default'))) {
            func(mockDefaultNode);
          } else if (String(regex) === String(RegExp('.a'))) {
            func(mockANode);
          } else if (String(regex) === String(RegExp('.b'))) {
            func(mockBNode);
          } else {
            throw new Error(`Invalid reg ex passed while walking the nodes: ${regex}`);
          }
        }),
    };

    plugin(mockRoot);

    expect(themeableVariableInformation).toMatchSnapshot();
  });
});
