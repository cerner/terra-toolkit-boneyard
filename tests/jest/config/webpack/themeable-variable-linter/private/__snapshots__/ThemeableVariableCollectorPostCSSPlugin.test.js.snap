// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Themeable Variable Collector Plugin populates the list of themeable variables 1`] = `
ThemeableVariableInformation {
  "perThemeVariableTracker": Object {
    "a": Object {},
    "b": Object {},
    "default": Object {},
  },
  "themes": Array [
    "default",
    "a",
    "b",
  ],
  "variables": Object {
    "--var1": Object {
      "name": "--var1",
      "origin": Object {
        "column": 1,
        "file": "file1",
        "line": 1,
      },
    },
    "--var2": Object {
      "name": "--var2",
      "origin": Object {
        "column": 1,
        "file": "file1",
        "line": 1,
      },
    },
  },
}
`;

exports[`Themeable Variable Collector Plugin populates the list of tracked variables 1`] = `
ThemeableVariableInformation {
  "perThemeVariableTracker": Object {
    "a": Object {
      "--var1": Array [
        Object {
          "name": "--var1",
          "origin": Object {
            "column": 1,
            "file": "file1",
            "line": 1,
          },
          "value": 1,
        },
      ],
    },
    "b": Object {
      "--var1": Array [
        Object {
          "name": "--var1",
          "origin": Object {
            "column": 1,
            "file": "file1",
            "line": 1,
          },
          "value": 1,
        },
      ],
      "--var2": Array [
        Object {
          "name": "--var2",
          "origin": Object {
            "column": 1,
            "file": "file1",
            "line": 1,
          },
          "value": 2,
        },
        Object {
          "name": "--var2",
          "origin": Object {
            "column": 1,
            "file": "file1",
            "line": 1,
          },
          "value": 3,
        },
      ],
      "--var3": Array [
        Object {
          "name": "--var3",
          "origin": Object {
            "column": 1,
            "file": "file1",
            "line": 1,
          },
          "value": 3,
        },
      ],
    },
    "default": Object {
      "--var1": Array [
        Object {
          "name": "--var1",
          "origin": Object {
            "column": 1,
            "file": "file1",
            "line": 1,
          },
          "value": 1,
        },
      ],
      "--var2": Array [
        Object {
          "name": "--var2",
          "origin": Object {
            "column": 1,
            "file": "file1",
            "line": 1,
          },
          "value": 2,
        },
      ],
      "--var3": Array [
        Object {
          "name": "--var3",
          "origin": Object {
            "column": 1,
            "file": "file1",
            "line": 1,
          },
          "value": 3,
        },
      ],
    },
  },
  "themes": Array [
    "default",
    "a",
    "b",
  ],
  "variables": Object {},
}
`;
